### PR TITLE
SLING-11119 - Optimise the service retrieval for bundled scripts

### PR DESCRIPTION
**Pull request by raducotescu**
* extended the `ServiceCache` with a `getServices` method that returns a filtered services array
* the `ExecutableUnits` are now backed by the `ServiceCache`

### CODE REVIEW

File: src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java [hunk 1]
```diff
@@ -101,83 +95,37 @@ public BundleContext getBundleContext() {
...
-                    }
-                    this.references.add(ref);
-                    this.services.put(className, result);
-                    return result;
-                }
-            }
+        try {
+            ClassLoader bundleClassloader = getBundle().adapt(BundleWiring.class).getClassLoader();
+            return (T) serviceCache.getService(bundleClassloader.loadClass(className));
+        } catch (ClassNotFoundException e) {
```
Comment (karlpauls) [inline src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java:110]: I guess this is quite a change so - previously, we would have allowed to lookup up services with interfaces that are not wired to the bundle. Now, that will fail - are we sure that nobody is using this to lookup services that are then used only via reflection (i.e., don't need the interfaces to be available)?

File: src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java [hunk 2]
```diff
@@ -101,83 +95,37 @@ public BundleContext getBundleContext() {
...
-                if (!objects.isEmpty()) {
-                    T[] srv = (T[]) Array.newInstance(bundle.loadClass(className), objects.size());
-                    result = objects.toArray(srv);
-                }
-            }
-        } catch (Exception e) {
-            LOG.error(String.format("Unable to retrieve the services of type %s.", className), e);
+            ClassLoader bundleClassloader = getBundle().adapt(BundleWiring.class).getClassLoader();
+            return (T[]) serviceCache.getServices(bundleClassloader.loadClass(className), filter);
+        } catch (ClassNotFoundException e) {
```
Comment (karlpauls) [inline src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java:123]: Same question here I guess: are we sure we don't have cases where somebody gets the service to use it via reflection?

File: src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java [hunk 3]
```diff
@@ -63,70 +70,151 @@ public void dispose() {
...
+                    Collections.sort(localReferences);
+                    Collections.reverse(localReferences);
+                    for (ServiceReference<ServiceType> ref : localReferences) {
+                        references.add(new Reference(ref));
+                    }
+                    synchronized (this) {
+                        SortedSet<Reference> existing = this.cache.get(key);
+                        if (existing != null) {
+                            existing.addAll(
+                                    references.stream().filter(reference -> !existing.contains(reference)).collect(Collectors.toList()));
```
Comment (karlpauls) [inline src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java:164]: I'm not sure the extra filter is needed - are we not already adding to a set?

File: src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java [hunk 4]
```diff
@@ -63,70 +70,151 @@ public void dispose() {
...
-                    this.bundleContext.ungetService(ref.reference);
+            }
+        }
+    }
+
+    private <ServiceType> SortedSet<Reference> getCachedReferences(Class<ServiceType> type) {
+        String key = type.getName();
+        SortedSet<Reference> references = cache.get(key);
+        if (references == null) {
+            references = Collections.synchronizedSortedSet(new TreeSet<>((o1, o2) -> -1 * o1.compareTo(o2)));
```
Comment (karlpauls) [inline src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java:150]: Doubt it matters much but should we maybe use a ConcurrentSkipListSet?

File: src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java [hunk 5]
```diff
@@ -63,70 +70,151 @@ public void dispose() {
...
-                    reference = existing;
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <ServiceType> ServiceType[] getServices(Class<ServiceType> type, String filter) {
+        List<ServiceType> result = new ArrayList<>();
+        try {
+            SortedSet<Reference> cachedReferences = getCachedReferences(type);
+            final Collection<ServiceReference<ServiceType>> filteredReferences = this.bundleContext.getServiceReferences(type, filter);
+            if (!filteredReferences.isEmpty()) {
+                List<ServiceReference<ServiceType>> localFilteredReferences = new ArrayList<>(filteredReferences);
```
Comment (karlpauls) [inline src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java:93]: I guess the extra list isn't really needed, right? We could just use Arrays.sort with reverse order for the sorting and the for loop already works with arrays (not that it matters much).

**Review by karlpauls — COMMENTED** [review]
LGTM overall - I had one or two minor suggestions but they might not matter. However, I have one question about the change to not be able to lookup services anymore when the interface is not available via the bundle class loader - maybe that is fine but I wanted to make sure we are not overlooking something?

File: src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java (hunk 4 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java:150]: For the comparator, you could use [`Comparator.reverseOrder()`](https://docs.oracle.com/javase/8/docs/api/java/util/Comparator.html#reverseOrder--) instead. IMHO it's more readable.

File: src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java (hunk 3 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java:164]: +1

File: src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java (hunk 1 above)
Comment (raducotescu) [reply src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java:110]: In HTL we always made sure that the object to load was accessible to the classloader with which the script was loaded. This was not necessarily the case for JSPs, but for bundled scripts they worked like that from the beginning - see https://github.com/apache/sling-org-apache-sling-scripting-spi/blob/47949fedc62380822961d8ef649130dee68a28e8/src/main/java/org/apache/sling/scripting/spi/bundle/BundledRenderUnit.java#L83.

File: src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java (hunk 2 above)
Comment (raducotescu) [reply src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java:123]: Same as above.

**Review by karlpauls — APPROVED** [review]

File: src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java (hunk 5 above)
Comment (raducotescu) [reply src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java:93]: We do need this list, as we need to sort the `ServiceReferences` according to their ranking, in order to be able to return the array with the `ServiceTypes`, the latter being objects that don't implement `Comparable`.

File: src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java (hunk 3 above)
Comment (raducotescu) [reply src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java]: Got rid of this in [712b011](https://github.com/apache/sling-org-apache-sling-scripting-core/pull/13/commits/712b0115040cf239d59eeb8baad13a175ff17309).

File: src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java (hunk 4 above)
Comment (raducotescu) [reply src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java]: Got rid of this in [712b011](https://github.com/apache/sling-org-apache-sling-scripting-core/pull/13/commits/712b0115040cf239d59eeb8baad13a175ff17309).

**Pull request merged**